### DISCUSSION TITLE
Cosmos DB - wykorzystywanie w firmach

### DISCUSSION THREAD

**Discussion by rmaziarka** [AzureCommunityPL/group#373] [Ogólna]
Siema grupo 😀

Mam ogólne pytanie o wasze doświadczenie wykorzystywania Cosmos DB w firmach z którymi współpracujecie / współpracowywaliście. Chciałem się dowiedzieć czy 
- aktualnie używacie
- używaliście i zmigrowaliscie na coś innego
- używaliście czegoś innego i zmigrowaliście na Cosmos DB

i jak wam się sprawdza, na jakie problemy się natknęliście i jak z nich wyszliście.

Mam mieszane uczucia bo sam ostatnimi czasy pisałem i utrzymywałem średniej wielkości rozwiązanie o CosmosDB - z dużą ilością pracy nad zrozumieniem ścieżek read/write, odpowiednim modelowania, denormalizacji to naprawdę dawało radę i utrzymywało się RU w ryzach. 

Jednoczeście jest mało materiałów jak skalować rozwiązania oparte na Cosmos DB. Masa materiałów, gdzie nawet ludzie z MS robią query po wszystkich partycjach. Brak informacji jak ludzie sobie radzili ze zmianami klucza partycji dla dużych aplikacji i baz. Nie ma jakichś bardziej zaawansowanych przykładów pokazujących update / migrację schemy. Wszystko ma niski entry level, a wokół dużo narzekania na Cosmos DB o wysokie RU i koszta.

Zastanawiam się nad sensem rozpoczynania kolejnych projektów na Cosmos DB bo w przypadku zespołów mających niewielkie doświadczenie w tej bazie to nawet nie ma materiałów by tych ludzi uczyć.

**Comment by mmisztal1980** [comment]
Mój kolega zrobił ostatnio b. fajną prezentację o tym jak korzystamy z CosmosDB, myślę że odpowie na przynajmniej część Twoich pytań
https://youtu.be/TiLqe29toTI?t=2781

**Reply by rmaziarka** [reply]
Ta prezentacja była bardziej nastawiona na jedną część problemu wydajności w CosmosDB - jak mierzyć wykorzystywane RU i się skalować. Fajna końcówka Kamila.

**Comment by ptrstpp950** [comment]
To ja wypowiem się za siebie z perspektywy 2 firm i kilku małych konsultacji:
Po pierwsze używam, używałem i będę używał oraz polecał użycie. Dlaczego?
- Cosmos DB jest szybki i ciężko w Azure znaleźć coś równie szybkiego
- dodatkowo rozwiązuje kilka istotnych problemów, których żadna inna baza w Azure nie daje (np dostęp mulit-write w wielu regionach). Co umożliwia rozwiązanie np: problemy wysokiej dostępności z punktu widzenia nie tylko availability zone, ale i regionów.
- niestety ma też wady np: do niedawna problem z backup
- do tego jest problem z dev, którzy posługują się sprawnie NoSQL, a nie próbują kopiować idee z SQL do NoSQL, albo nie daj Boże używać EF czy innego ORM'a
- do tego dochodzą kwestie braku zrozumienia cennika i to nie tylko wśród "technicznych" ale też "byłych technicznych" - Pewnie jeszcze długo będę opowiadał anegdotę jak to "Krzysztof Jarzyna ze Szczecina" zrobił "przyjemną rozmowę" po wprowadzeniu auto-scaling, który w tym wypadku był tańszy niż manual scale, że nie zachowujemy się jak inżynierowie tylko wprowadzamy rozwiązania bez zrozumienia, itd.
- Cennik raz jeszcze bo kilkukrotnie miał (r)ewolucje. Raz o mało co nie dokonaliśmy migracji systemu z Cosmos do SQL ale konferencja MS zmieniła naszą decyzję.
- jako bonus dochodzą problemy też wieku dziecięcego i na przykład raz zdarzyła się sytuacja, w której indeks nie działał bez hack

Co do materiałów. niestety to prawda. Może powinienem kurs zrobić, skoro już kilka rzeczy utopiłem w Cosmos ;)

**Comment by MichalJankowskii** [comment]
Zgodzę się z @ptrstpp950.

No może z jednym punktem bym dyskutował - dla mnie backup nie został do końca rozwiązany. 

Cytująć:
"If you have accidentally deleted or corrupted your data, before you create a support request to restore the data, make sure to increase the backup retention for your account to at least seven days. It’s best to increase your retention within 8 hours of this event. This way, the Azure Cosmos DB team has enough time to restore your account."

:)

Natomiast to co nikt wcześniej nie podkreślił - jeśli ktoś chce używać Cosmos zamiast relacyjnej bazy danych to nie jest to najlepszy case dla tego produktu. A to mi się zdarzało widzieć i potem pojawiały się potwory / rozgorzyczenie. Tu niestety trzeba przejść na myślenie obiektowe - ma to swoje wady i zalety.

A z drugiej strony jeśli potrzebujemy samą bazę obiektową to raczej sugerowałbym się zastanowić nad MongoDB.

Dobra, a kiedy CosmosDB - przykładowo gdy:
- generalnie jesli potrzebujemy wydajność
--- podgrupą problemów jest np. cache / read-model
- budujemy rozwiazanie o zasięgu "światowym"
- IoT - baza do przechowywania danych z ustawionym TTL

Oczywiście można użyć dla mniejszych problemów, ale wtedy nie jest to bardzo często najtańsze rozwiązanie.
